cpp/include/cuvs/neighbors/sample_filter.hpp

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/core/bitset.hpp>
+#include <raft/core/detail/macros.hpp>
+
+#include <cstddef>
+#include <cstdint>
+
+namespace cuvs::neighbors::filtering {
+
+/* A filter that filters nothing. This is the default behavior. */
+struct none_ivf_sample_filter {
+  inline _RAFT_HOST_DEVICE bool operator()(
+    // query index
+    const uint32_t query_ix,
+    // the current inverted list index
+    const uint32_t cluster_ix,
+    // the index of the current sample inside the current inverted list
+    const uint32_t sample_ix) const
+  {
+    return true;
+  }
+};
+
+/* A filter that filters nothing. This is the default behavior. */
+struct none_cagra_sample_filter {
+  inline _RAFT_HOST_DEVICE bool operator()(
+    // query index
+    const uint32_t query_ix,
+    // the index of the current sample
+    const uint32_t sample_ix) const
+  {
+    return true;
+  }
+};
+
+template <typename filter_t, typename = void>
+struct takes_three_args : std::false_type {};
+template <typename filter_t>
+struct takes_three_args<
```

Comment:
Definitino in `include/`, implementation compiled into source file(s) in `src/`.